core/src/main/java/jenkins/scm/RunWithSCM.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2014 Jesse Glick.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm;
+
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.User;
+import hudson.scm.ChangeLogSet;
+import hudson.util.AdaptedIterator;
+import jenkins.util.SystemProperties;
+import org.kohsuke.stapler.export.Exported;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.AbstractSet;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * @since FIXME
+ */
+public interface RunWithSCM<JobT extends Job<JobT, RunT> & Queue.Task,
+        RunT extends Run<JobT, RunT> & RunWithSCM<JobT,RunT> & Queue.Executable> {
+
+    List<ChangeLogSet<? extends ChangeLogSet.Entry>> getChangeSets();
+
+    @CheckForNull
+    Set<String> getCulpritIds();
+
+    boolean shouldCalculateCulprits();
```

Comment:
Some Javadoc needed for these. It is not really obvious from looking at source code what the relationship among the four culprit-related methods is.